torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3004,6 +3005,164 @@ struct to_ir {
+            // be a string, but we'll need to inspect the kwarg to find
+            // the type of its `value`
+            Expr value = apply.attributes()[0].value();
+            TypePtr value_type;
+            // Special case: we have a TK_CONST and need to
+            // differentiate between IntType and FloatType
+            if (value.kind() == TK_CONST) {
+              Const number_type = Const(value);
+              if (number_type.isFloatingPoint()) {
+                value_type = FloatType::get();
+              } else {
+                value_type = IntType::get();
+              }
+            } else {
+              value_type = emitSimpleExpr(value)->type();
+            }
+            type = DictType::create(StringType::get(), value_type);
+            if (type_hint && type != type_hint) {
+              throw ErrorReport(apply.range())
+                  << "Expected " << type->repr_str()
+                  << " type annotation for dict(), found "
+                  << type_hint->repr_str();
+            }
+          }
+          /// Otherwise, use the default of Dict[str, Tensor]. If we have an
+          // iterable in apply.inputs(), then this default type will be
+          // overridden when the iterable is emitted
+          else {
+            type = DictType::create(StringType::get(), TensorType::get());
+          }
+        }
+
+        // If we haven't been passsed an iterable/mapping, we have
+        // either an empty call to `dict()` or we have kwargs
+        if (apply.inputs().empty()) {
+          std::vector<Value*> keys;
+          std::vector<Value*> values;
+          for (auto attr : apply.attributes()) {
+            auto sl = StringLiteral::create(apply.range(), attr.name().name());
+            keys.push_back(emitStringLiteral(sl));
+            values.push_back(emitExpr(attr.value()));
+          }
+          return std::make_shared<SimpleValue>(
+              graph
+                  ->insertNode(graph->createDict(
+                      type->expect<DictType>()->getKeyType(),
+                      type->expect<DictType>()->getValueType(),
+                      keys,
+                      values))
+                  ->output());
+        }
+
+        // Otherwise, we need to emit a new dict using the
+        // iterable/mapping we've been passed (and potentially kwargs
+        // as well)
+        else {
+          // We can't feasibly register all possible key x value
+          // combinations of new prim ops for the case that we use the
+          // constructor with a dict literal. It makes much more sense
+          // to transform the dict literal into a list of tuples so that
+          // we can use the existing constructors
+          if (!apply.inputs().empty() &&
+              apply.inputs()[0].kind() == TK_DICT_LITERAL) {
+            auto dict_lit = DictLiteral(apply.inputs()[0]);
+            std::vector<Expr> zipped;
```

Comment:
Nit:
Add assert to make sure key_inputs() and value_inputs() have same size.

Also, slightly better to call `zipped.reserve(key_inputs().size())` to save some memory allocation time.